External/FEXCore/Source/Interface/Core/ObjectCache/JobHandling.cpp

```diff
@@ -13,11 +13,112 @@
 
 namespace FEXCore::CodeSerialize {
   void AsyncJobHandler::AsyncAddNamedRegionJob(uintptr_t Base, uintptr_t Size, uintptr_t Offset, const std::string &filename) {
-    // XXX: Actually add the add named region job
+    // This function adds a named region *JOB* to our named region handler
+    // This needs to be as fast as possible to keep out of the way of the JIT
+
+    auto BaseFilename = std::filesystem::path(filename).filename().string();
+
+    if (!BaseFilename.empty()) {
+      // Create a new entry that once set up will be put in to our section object map
+      auto Entry = std::make_unique<CodeRegionEntry>(
+        Base,
+        Size,
+        Offset,
+        filename,
+        NamedRegionHandler->DefaultCodeHeader(Base, Offset)
+      );
+
+      // Lock the job ref counter so we can block anything attempting to use the entry before it is loaded
+      Entry->NamedJobRefCountMutex.lock();
+
+      CodeRegionMapType::iterator EntryIterator;
+      {
+        std::unique_lock lk {CodeObjectCacheService->GetEntryMapMutex()};
+
+        auto &EntryMap = CodeObjectCacheService->GetEntryMap();
+
+        auto it = EntryMap.emplace(Base, std::move(Entry));
```

Comment:
The `!it.second` case can be slightly sped up by changing this to `EntryMap.emplace(Base, nullptr)`. The heap-allocating `std::make_unique<CodeRegionEntry>` only needs to be done in the `it.second == true` case then, which would explicitly overwrite the dummy entry created. (This assumes we can bear holding the lock a little bit longer in that case).

Reply:
I'm not sure I'm following what you mean here. We're not inserting a dummy entry. We're inserting an entry with the minimal amount of data required to kick off the async job.